cloudinit/distros/alpine.py

```diff
@@ -202,6 +204,423 @@ def preferred_ntp_clients(self):
+        # inputs. If something goes wrong, we can end up with a system
+        # that nobody can login to.
+        if use_useradd:
+            # Shadow package is installed, use 'useradd'
+            add_cmd = ["useradd", name]
+            log_add_cmd = ["useradd", name]
+
+            add_opts = {
+                "expiredate": "--expiredate",
+                "gecos": "--comment",
+                "groups": "--groups",
+                "homedir": "--home",
+                "inactive": "--inactive",
+                "passwd": "--password",
+                "primary_group": "--gid",
+                "shell": "--shell",
+                "uid": "--uid",
+            }
+
+            add_flags = {
+                "no_log_init": "--no-log-init",
+                "no_user_group": "--no-user-group",
+                "system": "--system",
+            }
+        else:
+            # Shadow package is not installed, use Busybox 'adduser'
+            add_cmd = ["adduser", "-D"]
+
+            add_opts = {
+                "expiredate": "",
+                "gecos": "-g",
+                "groups": "",
+                "homedir": "-h",
+                "inactive": "",
+                "passwd": "",
+                "primary_group": "-G",
+                "shell": "-s",
+                "uid": "-u",
+            }
+
+            add_flags = {"system": "-S"}
+
+        if use_useradd:
+            redact_opts = ["passwd"]
+
+        # support kwargs having groups=[list] or groups="g1,g2"
+        groups = kwargs.get("groups")
+        if groups:
+            if isinstance(groups, str):
+                groups = groups.split(",")
+
+            if isinstance(groups, dict):
+                LOG.warning(
+                    "DEPRECATED: The user %s has a 'groups' config value of"
+                    " type dict which is deprecated and will be removed in a"
+                    " future version of cloud-init. Use a comma-delimited"
+                    " string or array instead: group1,group2.",
+                    name,
+                )
+
+            # remove any white spaces in group names, most likely
+            # that came in as a string like: groups: group1, group2
+            groups = [g.strip() for g in groups]
+
+            # kwargs.items loop below wants a comma delimeted string
+            # that can go right through to the command.
+            kwargs["groups"] = ",".join(groups)
+
+            primary_group = kwargs.get("primary_group")
+            if primary_group:
+                groups.append(primary_group)
+
+        if create_groups and groups:
+            for group in groups:
+                if not util.is_group(group):
+                    self.create_group(group)
+                    LOG.debug("created group '%s' for user '%s'", group, name)
+        if "uid" in kwargs.keys():
+            kwargs["uid"] = str(kwargs["uid"])
+
+        addn_groups = None
```

Comment:
Nit: I think this section could be simplified. Something like:
```python
unsupported_busybox_values = {
    "addn_groups": None,
    "expirydate": None,
    "inactive": None,
    "passwd": None
}
```
Then later something along the lines of:
```python
for key, val in sorted(kwargs.items()):
    if key in unsupported_busybox_values:
        unsupported_busybox_values[key] = val
```

Reply:
Done.